### DISCUSSION TITLE
Nixos installation

### DISCUSSION THREAD

**Discussion by musicformellons** [akinomyoga/ble.sh#425] [Q&A]
I see some [instructions how to install on Nixos](https://github.com/akinomyoga/ble.sh/wiki/Manual-A1-Installation#nix-nixpkgs) but both manual and using home-manager seem to be 'in development' AFAICT. What is the best way to install currently on Nixos and what features am I missing out on?

**Comment by akinomyoga** [comment]
I think the description of the manual installation works. It says `ToDo: Fix blesh-share`, but I think it was related to the discussion about what `blesh-share` should print (e.g. in https://github.com/NixOS/nixpkgs/pull/181963#discussion_r933612333), but it's already fixed.

On the other hand, I think the installation with Home Manager doesn't work now because it's not yet merged upstream (as it references https://github.com/nix-community/home-manager/pull/3238).

**Reply by akinomyoga** [reply]
@aiotter Do you have additional comments?

**Reply by aiotter** [reply]
You're right. Home-manager is a preferred way because blesh.sh stores configuration data in per-user basis, however, it is not merged yet because I have been extremely busy nowadays moving to a new residence.

**Reply by aiotter** [reply]
fyr, my .bashrc is like:
```
[[ $- == *i* ]] && source '/nix/store/jyry6zvchsspidk3ci9gqy0a9r6is4dn-blesh-0.4.0-devel3/share/blesh/ble.sh' --attach=none '--rcfile=/nix/store/qav8qcsii44r7k138dw5v85lw24083zd-blerc'

...

source /nix/store/7w9vbh9iwxllw9jarz3141vlifc23klj-bash-preexec-0.5.0/share/bash/bash-preexec.sh
[[ ${BLE_VERSION-} ]] && ble-attach
```

**Reply by akinomyoga** [reply]
Thanks for the information!

**Comment by musicformellons** [comment]
Thanks for your quick reply! I have:
```
  systemd.services.create-bashrc = {
    enable = true;
    description = "Create .bashrc file";
    script = ''
      set -e
      mkdir -p /home/${my.username}
      cat > /home/${my.username}/.bashrc << 'EOF'
      source "$(blesh-share)"/ble.sh --attach=none # does not work currently
      eval "$(atuin init bash)"
      export EDITOR=vim
      set -o vi
      [[ "''${BLE_VERSION-}" ]] && ble-attach
      EOF
      chown ${my.username}:users /home/${my.username}/.bashrc
    '';
    serviceConfig = {
      Type = "oneshot";
      RemainAfterExit = "yes";
    };
  };
  ```
  in my `configuration.nix`.
  In general ble.sh works as should, just one thing.
  On opening a new shell I get:
```
ble.sh: XDG_RUNTIME_DIR='/run/user/501' is not a directory.
bleopt: option `keymap_vi_mode_string_nmap' not found
ble.sh: XDG_RUNTIME_DIR='/run/user/501' is not a directory.
bleopt: option `keymap_vi_mode_string_nmap' not found
```
before i get the normal prompt. BTW: No idea whether relevant, but I have `xdg-utils` installed and also `direnv` and `nix-direnv`.
I tried to set that normal mode indicator by having `~/.blerc` with:
```
# in blehook keymap_vi_load
bleopt keymap_vi_mode_string_nmap=$'\e[1m-- NORMAL --\e[m'
```
It's not picked up as in normal mode I still get '~' instead of '-- NORMAL __'.

**Reply by akinomyoga** [reply]
- **Q1**: What are the results of the following commands?

```bash
$ echo "$XDG_RUNTIME_DIR"
$ ls -la "$XDG_RUNTIME_DIR"
```

**Reply by musicformellons** [reply]
```
pq in 🌐 devbox in gb on  main [!] 
⬢ [Systemd] ❯ echo "$XDG_RUNTIME_DIR"
/run/user/501

pq in 🌐 devbox in gb on  main [!] took 43s 
⬢ [Systemd] ❯ ls -la "$XDG_RUNTIME_DIR"
drwx------ - pq 16 Mar 15:02 blesh
srw-rw-rw- - pq 16 Mar 15:02 bus
drwxr-xr-x - pq 16 Mar 15:02 systemd
srwxr-xr-x - pq 16 Mar 15:02 vscode-git-a70d2791ed.sock
srwxr-xr-x - pq 16 Mar 15:02 vscode-ipc-45c81983-a9b0-4318-b0e0-2740738f1f6d.sock
srwxr-xr-x - pq 16 Mar 15:02 vscode-ipc-60f4df65-f9de-4f82-8047-c9d4e7d36f61.sock
srwxr-xr-x - pq 16 Mar 15:02 vscode-ipc-2515f7a0-9aa8-442f-a75d-56d89c60d99a.sock
srwxr-xr-x - pq 16 Mar 15:02 vscode-ipc-36709e57-b952-4fef-b39a-ec1e1fb03166.sock
srwxr-xr-x - pq 16 Mar 15:02 vscode-ipc-57437f35-7070-4b89-8b8c-fb9cdb074733.sock
srwxr-xr-x - pq 16 Mar 15:02 vscode-ipc-a0cbaa9d-5611-4036-bf4c-6b9e0c07ad04.sock
srwxr-xr-x - pq 16 Mar 15:02 vscode-ipc-af0f1720-3f71-4021-acf6-12b72234fd9a.sock
srwxr-xr-x - pq 16 Mar 15:02 vscode-ipc-d7bc9e0f-8bc8-4ffa-b032-084f4315abe9.sock
srwxr-xr-x - pq 16 Mar 15:02 vscode-ipc-e3d9a24a-2f2f-499b-863b-60b11e5a718f.sock
srwxr-xr-x - pq 16 Mar 15:02 vscode-ipc-f966b2d0-ec2d-41b4-97f6-f4f10a99014a.sock
lrwxrwxrwx - pq 16 Mar 15:02 vscode-ssh-auth-sock-322542710 -> /opt/orbstack-guest/run/host-ssh-agent.sock
```
I'm working on a nixos box via orbstack and vscode remote-ssh extension.

My code/repo is local (mac) and uses orbstack filesharing with the nixos box.

**Reply by akinomyoga** [reply]
> I tried to set that normal mode indicator by having `~/.blerc` with:
> 
> ```
> # in blehook keymap_vi_load
> bleopt keymap_vi_mode_string_nmap=$'\e[1m-- NORMAL --\e[m'
> ```

**Q2**: Where do you put the above line?

You need to put it in `blehook keymap_vi_load`.

**Reply by akinomyoga** [reply]
```
ble.sh: XDG_RUNTIME_DIR='/run/user/501' is not a directory.
bleopt: option `keymap_vi_mode_string_nmap' not found
ble.sh: XDG_RUNTIME_DIR='/run/user/501' is not a directory.
bleopt: option `keymap_vi_mode_string_nmap' not found
```

Do you source ble.sh twice in the same Bash session?

- **Q3**: What is the output when you put the following lines in `~/.blerc`?

```bash
# blerc

declare -p BLE_SESSION >&2
ble-stackdump >&2
```

**Reply by musicformellons** [reply]
Q2: I see... I did it wrong then (just had these two lines in a ~/.blerc file). But then I need to alter probably an existing .blerc file...?! where?
Q3: Looks indeed like it's sourced twice (i'm new to nix/nixos and this particular setup with repo local and using filesharing...), don't know why. What ~/.blerc do you mean, as I now created that myself with just 2 lines (I thought that might overwrite default settings).

**Reply by musicformellons** [reply]
I'm now deriving a ~/.blerc from the blerc.template in the ble.sh-repo, see if that works. Sorry for my ignorance.

**Reply by musicformellons** [reply]
I restarted my laptop. It's ok now (with a proper ~/.blerc file). Thanks! Very nice!

**Reply by musicformellons** [reply]
Oh..., after `sudo nixos-rebuild switch` the `ble.sh: XDG_RUNTIME_DIR='/run/user/501' is not a directory.` returns...

**Reply by musicformellons** [reply]
When adding the suggested lines to ~/.blerc I get:
```
bash: declare: BLE_SESSION: not found
```

**Reply by akinomyoga** [reply]
Ah, sorry. It is my typo. Could you try the following?

```bash
# blerc

declare -p BLE_SESSION_ID >&2
ble-stackdump >&2
```

**Reply by musicformellons** [reply]
```
declare -- BLE_SESSION_ID="1710738954.006495/2835"
declare -- BLE_SESSION_ID="1710738954.006495/2835"
```

**Reply by akinomyoga** [reply]
Thanks! It means that `~/.bashrc` is sourced in the same session (but not in a temporary session).

Hmm, `ble-stackdump` is not enabled. Could you try the following?

```bash
# blerc

declare -p BLE_SESSION_ID >&2
bleopt internal_stackdump_enabled=1
ble-stackdump >&2
```

**Comment by kaBeech** [comment]
Oh! I actually added to the Wiki about this last night without noticing this discussion. Here's what works for me. Note that I use Flake notation throughout, though it will probably also work without Flakes:

## NixOS

```nix
# configuration.nix
{pkgs, ...}: {
  environment.systemPackages = with pkgs; [
    blesh
  ];
}
```

Note that I didn't use the `programs.bash.blesh.enable` strategy in `configuration.nix` because when I tried that it didn't add `blesh-share` to my PATH and I didn't feel like troubleshooting that at the moment.

`blesh` package provides `blesh-share` executable, which echoes the directory path in which `ble.sh` is installed.
You can use it to configure ble.sh in your `~/.bashrc`.

```bash
# ~/.bashrc
source "$(blesh-share)"/ble.sh --attach=none # does not work currently (this comment was here when I arrived and, for Chesterton's fence reasons, I have not removed it. I have not noticed anything not working)
...
[[ ! ${BLE_VERSION-} ]] || ble-attach
```

## Home Manager

```nix
# home.nix
{pkgs, ...}: {
  home.packages = with pkgs; [
    blesh
  ];
  programs.bash = {
    enable = true;
    bashrcExtra = ''
      [[ $- == *i* ]] && source "$(blesh-share)"/ble.sh --noattach
      ...
      [[ ! ${BLE_VERSION-} ]] || ble-attach
    '';
  };
}
```

Hope this helps!

**Comment by Peter-Flynn** [comment]
I ended up rewriting the above home-manager solution, as it didn't work for me out-of-the-box. Posting here in case it helps someone else:
```                                                                                                   
{ pkgs, lib, ... }:

{
  home.packages = with pkgs; [ blesh ];
  programs.bash = {
    enable = true;
    initExtra = lib.mkMerge [
      (lib.mkBefore ''
        [[ $- == *i* ]] && source "$(blesh-share)"/ble.sh --noattach
      '')
      (lib.mkAfter ''
        [[ ! ''\${BLE_VERSION-} ]] || ble-attach
      '')
    ];
  };
}
```
